File: Algorithms/StringMatch/kmp.py
def genLps(pattern):
    lps = [0] * (len(pattern)+1)
    i = 1

    lpsLen = 0

    while i < len(pattern):
        if pattern[i] == pattern[lpsLen]:
            lpsLen += 1
            lps[i] = lpsLen
            i += 1
        else:
            if lpsLen != 0:
                lpsLen = lps[lpsLen - 1]
            else:
                lps[i] = 0
                i += 1

    return lps

def kmp(txt, pattern):
    lps = genLps(pattern)

    i = 0
    j = 0

    while i < len(txt):
        if pattern[j] == txt[i]:
            i += 1
            j += 1

        if  j == len(pattern):
            print(f"Found pattern at index {i-j}")
            j = lps[j-1]

        elif i < len(txt) and pattern[j] != txt[i]:
            if j != 0:
                j = lps[j-1]
            else:
                i += 1

kmp("ababdabacdababcabab", "ababcabab")

File: Algorithms/StringMatch/kmp.rs
fn kmp(text: &str, pattern: &str) {
    let mut lps: Vec<usize> = Vec::with_capacity(pattern.len());

    for _ in 0..(pattern.len()) {
        lps.push(0);
    }

    let (mut prev, mut i) = (0 as usize, 1 as usize);
    
    while i < pattern.len() {
        if pattern.chars().nth(prev).unwrap() == pattern.chars().nth(i).unwrap() {
            lps[i] = lps[prev] + 1;
            prev += 1;
            i += 1;
        } else if prev == 0 {
            lps[i] = 0;
            i += 1
        }
        else {
            prev -= 1;
        }
    }

    println!("Text {text}");
    println!("Pattern {pattern}");
    println!("Longest Prefix Suffix Table: {lps:?}");

    let (mut i, mut j): (usize, usize) = (0, 1);
    let mut longest = 0;

    while i < text.len() {
        if text.chars().nth(i).unwrap() == pattern.chars().nth(j).unwrap() {
            (i, j) = (i + 1, j + 1);
        } else {
            if j == 1 {
                i += 1;
                j = 1;
            } else {
                j = lps[j - 1]
            }
        }

        if j >= pattern.len() {
            longest = i - text.len() as usize;
            break;
        }
    }

    println!("Text {text}");
    println!("Pattern {pattern}");
    println!("Longest {longest}");
}

fn main() {
    kmp("aacaaa", "aaa")
    // kmp("a", "aaacaaaa")
}
